src/product/product.service.ts

```diff
@@ -0,0 +1,82 @@
+import {
+  BadRequestException,
+  Injectable,
+  InternalServerErrorException,
+} from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { validate as isUUID } from 'uuid';
+
+import { CreateProductDto } from './dto';
+import { Product } from './entities/product.entity';
+import { PaginationDto } from '../common/dto/pagination.dto';
+
+@Injectable()
+export class ProductService {
+  constructor(
+    @InjectRepository(Product)
+    private readonly productRepository: Repository<Product>,
+  ) {}
+
+  async findOne(searchTerm: string) {
+    let product: Product;
+
+    if (isUUID(searchTerm)) {
+      product = await this.productRepository.findOneBy({ id: searchTerm });
+    } else {
+      product = await this.productRepository.findOneBy({ name: searchTerm });
+    }
+
+    if (!product) throw new BadRequestException('Product not found');
+
+    return product;
+  }
+
+  async findAll(paginationDto: PaginationDto) {
+    const { limit = 10, offset = 0, sort = 'DESC' } = paginationDto;
+
+    return await this.productRepository.find({
+      take: limit,
+      skip: offset,
+      order: {
+        createdAt: sort,
+      },
+    });
+  }
+
+  async create(createProductDto: CreateProductDto) {
+    try {
+      const product = this.productRepository.create(createProductDto);
+      return await this.productRepository.save(product);
+    } catch (error) {
+      this.handleDbError(error);
+    }
+  }
+
+  async update(id: string, updateProductDto: CreateProductDto) {
+    const product = await this.productRepository.preload({
+      id,
+      ...updateProductDto,
+    });
+
+    if (!product) throw new BadRequestException('Product not found');
+
+    try {
+      await this.productRepository.save(product);
+      return product;
+    } catch (error) {
+      this.handleDbError(error);
+    }
+  }
+
+  async remove(id: string) {
+    const product = await this.findOne(id);
+    await this.productRepository.remove(product);
+  }
+
+  private handleDbError(error: any) {
+    if (error.code === '23505') throw new BadRequestException(error.detail);
```

Comment:
Could you store this error code in a constant?

Reply:
sure